### PR TITLE
Optimiser and refactorings.

### PR DESCRIPTION
**Pull request by MattWindsor91**
This pull request adds in an **optimising stage** that eliminates large amounts of Boolean expression cruft by sending terms through various optimisation passes.  More passes could be added, but the existing ones seem to allow HSF models of the ticketed lock to proceed through `qarmc` properly.

I've also ‘optimised’ the **internal representation of model objects**, making things hopefully more consistent and precise.  Terms now have the fact that they only have one goal view baked into the type.  Final terms use only single-func views, and **the HSF view-to-func generator has been brought forwards**.

`Flattener` **is now** `Destructurer`; the new stages are `Flattener` (convert views to funcs) and `Optimiser` (optimise).  `HSF` is now merged with `Horn`, as we only have the one Horn clause generator!

There is also some work trying to **improve Starling's doc comments**, including an attempt to get Sandcastle to generate Starling documentation.  While I'm keeping this in tree because it's a useful way of visualising doc comments, I've unplugged it from the solution and am considering ditching it for something that actually works properly with F♯.  **Since it isn't in the solution, it doesn't add any build dependencies.**

### CODE REVIEW

File: Examples/Pass/ticketLockNonAtomicRelease2.cvf [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  {| holdLock() |}
+}
+
+/*
+ * Unlocks the ticket lock.
+ */
+method unlock() {
+  {| holdLock() |}
+    <s = serving>;
+  {| /*relLock(s)*/ holdLock() |}
```
Comment (mjp41) [inline Examples/Pass/ticketLockNonAtomicRelease2.cvf:31]: I don't see why this proof goes through?  s is not mentioned, so it seems it can take any value, and thus I don't see why the next command is sound.  Thoughts?  Is this a bug?

Comment (MattWindsor91) [inline Examples/Pass/ticketLockNonAtomicRelease2.cvf:31]: This looks sound to me, but for unexpected reasons.

`{| holdLock() |} <serving = (s + 1)> {| emp |}` looks ok, because `emp` is constrained to `true` in this version.  This is an odd choice for `emp`, but came from HSF.

Since we're not constraining the value of `serving` afterwards, we can get to various different situations, all of which seem sound:
- `ticket >= (s + 1) > serving`, which is expected behaviour and what we want it to be, morally.
- `(s + 1) == serving`, ie don't actually give up the lock at all.  Sound because we never actually state that we got rid of the lock, just that we forgot we had one.  This will deadlock (nobody who has a ticket can ever get the lock), but we're only showing partial correctness, so this goes through.
- `serving > ticket`, which is a surprising turn of affairs, but is also valid: nobody who has a ticket will ever be served, and locking will deadlock (the do-while loop will spin forever), but that's ok because we never explicitly rely on this.

The bottom line is that `serving <= ticket` in this case is actually a problem/liveness constraint, not a proof one, and without it HSF merrily proves programs that will deadlock but satisfy their specifications.

Comment (mjp41) [inline Examples/Pass/ticketLockNonAtomicRelease2.cvf:31]: Okay. That makes sense, it is basically a lost wake up, which doesn't contradict mutual exclusion.
